definitions/types/azure-resource-id.json

```
@@ -6,7 +6,7 @@
   "examples": [
     "/subscriptions/12345678-1234-1234-abcd-1234567890ab/resourceGroups/resource-group-name/providers/Microsoft.Network/virtualNetworks/network-name"
   ],
-  "pattern": "^\/subscriptions\/[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}\/resourceGroups\/[a-zA-Z0-9-_.()]{0,89}[a-zA-Z0-9-_()]\/providers\/Microsoft\\.[a-zA-Z]+\/[a-zA-Z]+\/[0-9A-Za-z-]+$",
+  "pattern": "^\/subscriptions\/[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}\/resource(g|G)roups\/[a-zA-Z0-9-_.()]{0,89}[a-zA-Z0-9-_()]\/providers\/Microsoft\\.[a-zA-Z]+\/[a-zA-Z]+(?:\/[a-zA-Z0-9-_.()]+)+$",
```

What's the goal of the change at the end here? This now allows things to end with periods, is that what we want?

---

This is a generic blanket regex that allows tons of different variations of resource IDs without erroring out. The resource ID will come from a provisioned resource. We can put restrictions on the naming of resources to make sure the name is valid. For now, we don't ask the user to specify a resource ID, we just grab the generated one.

---

Is it ever valid to have a resource id that ends with a period?

---

No, but resource IDs will be grabbed automatically for now. And in the future if we allow users to import their resource IDs into Massdriver, they'll be copy/pasting the ID anyway.

---

Yeah, we have to make it more permissive. Basically we are facing one of two options:
* Make a more permissive `azure-resource-id` that allows *all* resource IDs, even if is more permissive for some resources than others
* Make multiple restrictive `azure-whatever-resource-id` types that will specifically and accurately only allow resource-id's for that resource

I think option 1 is much better than 2 - it also matches what we've done for AWS ARN and GCP GRN.  That last block you mention is the block that says "any number of `/some-thing` blocks are allowed. It's the most "permissive" block since it needs to cover the union of naming conventions of all the different Azure services.  For the time being, all it really needs to be is `[a-zA-Z0-9-]` cuz that is our internal naming convention. But in the event we ever need to ask a user for a resource ID and we use this type, we should make sure it can accept the resource ID. And it will at least prevent users from putting in the resource name instead.

@mclacore I wasn't able to find any Azure docs that listed resource naming conventions across Azure services. I just copied what they allow for resource groups. That may or may not be accurate.

---

> @mclacore I wasn't able to find any Azure docs that listed resource naming conventions across Azure services. I just copied what they allow for resource groups. That may or may not be accurate.

What we have is accurate for the resources we currently provision (and might potentially provision in the future).

---

Yeah the GCP GRN was a barn fire for similar reasons. I had to back it out to pretty much `projects/[a-z]/` because it was difficult to cover networks, topics, etc etc.

Taking that approach here sounds good and I'll approve this to get out of the way. To refine my comment a bit, is this maybe a little better last capture group?

`(?:\/[a-zA-Z0-9-_.()]*[a-zA-Z0-9-_()]+)+`

Basically, zero to many of what you have, then one to many of that same group w/o a period. 🤷

---

^ if thats true for all Azure resources, we can add it. I just don't think we know yet.  As of now, 100% of the time RGs come from a resource we've provisioned, so ideally any naming convention issues were handled granularly in the params.  If we ever allow a user to specify a resource ID, I expect 99.9999% of the time they are gonna be copy-pasting it (nobody is gonna type out the full subscription ID). In all those cases it should be come from an existing resource, which means it should be valid. Which is why I'm not that concerned with having extremely tight validation here.

I mostly just want it to prevent people from putting things in that aren't resource-id's, like resource names, or resouce groups, etc.

---

i do like @WillBeebe 's suggestion!
```suggestion
  "pattern": "^\/subscriptions\/[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}\/resource(g|G)roups\/[a-zA-Z0-9-_.()]{0,89}[a-zA-Z0-9-_()]\/providers\/Microsoft\\.[a-zA-Z]+\/[a-zA-Z]+((?:\/[a-zA-Z0-9-_.()]*[a-zA-Z0-9-_()]+)+$",
```

---

> I expect 99.9999% of the time they are gonna be copy-pasting it

@chrisghill totally agree this will be the user pattern. however, periods are a really common copy / paste buffer issue for example imagine the slack message:
```
hey team, chris is going to look into the issue with /subscriptions/4718857a-5dbe-452c-a172-21a3ae74304f/resourceGroups/jake-azure-jakesandbox-hs9h/providers/Microsoft.Network/virtualNetworks/jake-azure-jakesandbox-hs9h/subnets/jake-databricks-dev-container.  We'll update this channel once resolved!
```
you might copy the period that got added to the end that was added by your colleague's keyboard adding periods after double space, or because they had another sentence in their message. 
I've personally definitely done this dozens of times in my own life copying id's from messengers and scratched my head for a sec.